timeseries/src/autogluon/timeseries/metrics/abstract.py

```diff
@@ -0,0 +1,165 @@
+from typing import List, Optional
+
+import numpy as np
+import pandas as pd
+
+from autogluon.timeseries import TimeSeriesDataFrame
+from autogluon.timeseries.utils.datetime import get_seasonality
+from autogluon.timeseries.utils.warning_filters import warning_filter
+
+
+class TimeSeriesScorer:
+    """Base class for all evaluation metrics used in AutoGluon-TimeSeries.
+
+    This object always returns the metric in greater-is-better format.
+
+    Follows the design of ``autogluon.core.metrics.Scorer``.
+
+    Attributes
+    ----------
+    greater_is_better : bool, default = False
+        Whether internal method :meth:`~autogluon.timeseries.metrics.TimeSeriesScorer.compute_metric` is
+        a loss function (default), meaning low is good, or a score function, meaning high is good.
+    optimum : float, default = 0.0
+        The best score achievable by the score function, i.e. maximum in case of scorer function and minimum in case of
+        loss function.
+    optimized_by_median : bool, default = False
+        Whether given point forecast metric is optimized by the median (if True) or expected value (if False). If True,
+        all models in AutoGluon-TimeSeries will attempt to paste median forecast into the "mean" column.
+    is_quantile_metric : bool, default = False
+        Whether the given metric uses the quantile predictions. Some models will modify the training procedure if they
+        are trained to optimize a quantile metric.
+    equivalent_tabular_regression_metric : str
+        Name of an equivalent metric used by AutoGluon-Tabular with ``problem_type="regression"``. Used by models that
+        train a TabularPredictor under the hood. This attribute should only be specified by point forecast metrics.
+    """
+
+    greater_is_better: bool = False
+    optimum: float = 0.0
+    optimized_by_median: bool = False
+    is_quantile_metric: bool = False
```

Comment:
Assuming by "users" you mean other code in AutoGluon that uses this property - it would definitely be possible to use `isinstance` to accomplish the same.

My motivation for exposing this attribute was to simplify the instructions for human AutoGluon users that want to implement a custom metric. Instructions accompanying the current design:
> Inherit from `TimeSeriesScorer`; set `is_quantile_metric=True` if the metrics uses the quantile forecast

Instructions if we rely on checking `isinstance`
> If your metrics uses quantile forecast, inherit from `QuantileForecastScorer`; if your metric uses point forecast, inherit from `PointForecastScorer`. 

I don't really have a strong preference here, do you think my justification warrants introducing a separate `is_quantile_metric` attribute?

Reply:
I think the latter is cleaner. There should be a single version of truth, preventing some future implementations from relying on the flag and others relying on the type; which could diverge.

Reply:
> Users want to define a metric that considers both mean and quantile forecasts

> in general, always uses attributes /properties rather than isinstance checks to convey such information.

then why is there a type for `QuantileForecastScorer`? Maybe they should all be just scorers.

Reply:
Got rid of the types for `QuantileForecastScorer` and `PointForecastScorer` - the logic is now contained in helper methods `_get_quantile_forecast_score_inputs` and `_get_point_forecast_score_inputs`. Now only only the attribute `needs_quantile` carries the information of whether the metric uses the quantiles.